Non standard http

spirit-1984 opened the pull request:
See #1919 for details

spirit-1984 commented:
The thing that actually bothers me is the performance issue. Especailly considering the potential creation of many io.netty...HttpMethod instances for the same non-standard method. For now I just solved the problem using concurrent hash map. But maybe DefaultRouteBuilder should analyze all available methods and create this map at start. I am not sure how to pass that singleton map properly to the request.

graemerocher commented:
@spirit-1984 Thanks for the PR, will take a look soon. This change may be too big to go into a point release of 1.2.x. We will discuss that shortly. Thanks.

jameskleeh commented on http-client/src/main/java/io/micronaut/http/client/NettyClientHttpRequest.java:
```diff
@@ -180,20 +205,35 @@ protected QueryStringDecoder createDecoder(String uri) {
+     * @see @link https://bugs.openjdk.java.net/browse/JDK-8161372
+     */
+    private static io.netty.handler.codec.http.HttpMethod getMethod(String httpMethodName) {
+        if (!METHOD_MAP.containsKey(httpMethodName)) {
```
This method isn't thread safe

spirit-1984 replied:
You mean ConcurrentHashMap.containsKey is not thread safe? I assumed it's free of locks, unlike computeIfAbsent in case of the key actually present. So, my idea was:

- At startup different threads can block each other for a while when calling computeIfAbsent - but it thread safe, it will not produce inconsistent results.
- After all the possible http methods run at least once, all threads will just call containsKey and get for METHOD_MAP, and these methods are lock free and don't change anything.

Well, I can do this in another way - by modifying DefaultRouteBuilder which actually stores the possible routes and therefore knows all the methods that are available. And he can provide a map to another singleton bean, I guess, which can be used here. All the standard http methods should be in the map regardless of the actual routes, so that would not change the behaviour of the server for them.

spirit-1984 replied:
Let me clarify a little bit more what we are doing here. Basically the most obvious choice is to compute everything directly:

```java
return METHOD_MAP.computeIfAbsent(httpMethodName, io.netty.handler.codec.http.HttpMethod::valueOf);
```

Seems like pretty straightforward. But there is a catch here - in Java 8 this method puts a lock even if the key is present in the map (they fixed that in 9). So multiple requests will create a bottleneck. So I did it like this just so that people using Micronaut with java 8 would not get sudden performance issues. containsKey does not create any locks. If it returns false, we call computeIfAbsent. This method guarantees atomicity of read and write in case of absence, so two separate threads will get the same instance of Netty HtpMethod class. If containsKey returns true, the following call of get will return the non-null value also without locks.

I can rewrite it like this:

```java
private static io.netty.handler.codec.http.HttpMethod getMethod(String httpMethodName) {
        io.netty.handler.codec.http.HttpMethod value = METHOD_MAP.get(httpMethodName);
        if (value == null) {
            return METHOD_MAP.computeIfAbsent(httpMethodName, io.netty.handler.codec.http.HttpMethod::valueOf);
        }

        return value;
    }
```

That implementation should have no heavy side effects for read and should not introduce any inconsistencies with computeWithAbsent

spirit-1984 replied:
Of course, the simplest variant would be not to use any map like at all. But this could cause performance issues for high load, since netty HttpMethod.valueOf method returns the same instance for standard methods ,but will create a new instance otherwise for each call. That means, if you have a million REPORT requests, you'll get a million of HttpMethod instances. Of course, GC will get rid of them eventually, but I'm worried, it still can cause bottlenecks.

jameskleeh commented on http/src/main/java/io/micronaut/http/HttpMethod.java:
```diff
@@ -100,8 +101,23 @@ public static boolean requiresRequestBody(HttpMethod method) {
+        return method != null && (requiresRequestBody(method)
+                || method.equals(OPTIONS)
+                || method.equals(DELETE)
+                || method.equals(CUSTOM)
```
Why should we assume the method CUSTOM permits a body?

spirit-1984 replied:
Well, I am not sure about that. Some methods definitely do that (for example, REPORT in webdav protocol), so we cannot state otherwise. I can provide СustomHttpMethod with additional property body with values NONE|PERMITS|REQUIRES, but this is definitely demanding a much serious rewrite of the checks. If I drop CUSTOM from this method check, it would still be possible to send body for it, just like for GET, for example?

jameskleeh commented on http/src/main/java/io/micronaut/http/HttpResponse.java:
```diff
@@ -174,7 +174,7 @@ default String reason() {
      * @return The response
      */
-    static <T> MutableHttpResponse<T> notAllowed(Set<HttpMethod> allowed) {
+    static <T> MutableHttpResponse<T> notAllowed(Set<? extends CharSequence> allowed) {
```
This is a breaking change

spirit-1984 replied:
My mistake, I'll fix that

jameskleeh commented on http/src/main/java/io/micronaut/http/MutableHttpHeaders.java:
```diff
@@ -193,7 +193,7 @@ default MutableHttpHeaders auth(String userInfo) {
      * @return This HTTP headers
      */
-    default MutableHttpHeaders allow(Collection<HttpMethod> methods) {
+    default MutableHttpHeaders allow(Collection<? extends CharSequence> methods) {
```
This is a breaking change

spirit-1984 replied:
My mistake, I'll fix that

jameskleeh commented on router/src/main/java/io/micronaut/web/router/CustomMethodRoute.java:
```diff
@@ -0,0 +1,235 @@
+ * @author spirit-1984
+ * @since 1.2.1
+ */
+class CustomMethodRoute implements UriRoute {
```
Why is this class necessary?

spirit-1984 replied:
Ok, so I did this intentionally, cause I did not want to mess around with internals of DefaultUriRoute. The logic is that this class is used only for non-standard http methods, and the standard are working with the same code of DefaultUriRoute. I agree, this is overengineering, and too much worry. I'll fix that and modify DefaultUriRoute

spirit-1984 commented:
@jameskleeh I fixed all the problems, except ConcurrentHashMap usage. This needs further discussion.

spirit-1984 commented:
@jameskleeh  OK, I've changed my mind and removed ConcurrentHashMap. I guess modern GC implementations will handle the creation of instances smoothly.

spirit-1984 commented:
@jameskleeh What seems to be the issue now about thread safety?

jameskleeh commented:
> @jameskleeh What seems to be the issue now about thread safety?

You said you removed the caching of the instances which according to you:

> That means, if you have a million REPORT requests, you'll get a million of HttpMethod instances

I don't think that is a good idea to allow that to happen

spirit-1984 commented:
> I don't think that is a good idea to allow that to happen
Well modern GC should actually handle that pretty smooth. I'll investigate a little bit further.

jameskleeh commented:
` METHOD_MAP.computeIfAbsent` should be used. If there is a potential performance issue with that in JDK 8, then I would advise users to not use JDK 8

spirit-1984 commented:
> ` METHOD_MAP.computeIfAbsent` should be used. If there is a potential performance issue with that in JDK 8, then I would advise users to not use JDK 8

OK, thanks, I'll do that. That actually is a lot easier, than my dirty hack specifically for JDK 8

Im-dex commented:
Adding ConcurrentHashMap for HttpMethod instances caching decreases performance for the standard http methods, because you add redundant synchronization in that case. Since non-standard methods usage is a rare case, maybe it will be better to continue using netty cache for standard methods and create new instances otherwise to keep current performance?

spirit-1984 commented:
I agree with Im-dex about that, since Netty HttpMethod.valueOf actually returns cached instances for standard methods, therefore it will not hurt performance. For those that do want to use non-standard methods it would be required to use Java newer than 8, that have good GC capable of dealing with such intensive instance creation

graemerocher commented:
@spirit-1984 Looking to get this included shortly, would you mind resolving the conflicts?

spirit-1984 commented:
@graemerocher I did that but the master is unstable right now, have to wait for it to become stable, since the tests do not pass

spirit-1984 commented:
@graemerocher  OK, some tests failed after resolving conflicts. Will look to that tonight

spirit-1984 commented:
@graemerocher  Fixed all the conflicts

graemerocher commented:
@spirit-1984 Thanks for the contribution. Nice work!
